editor/user_stats.cpp

```diff
@@ -60,9 +89,100 @@ bool UserStats::Update()
     return false;
   }
 
-  m_changesCount = document.select_node("mmwatch/edits/@value").attribute().as_int(kUninitialized);
-  m_rank = document.select_node("mmwatch/rank/@value").attribute().as_int(kUninitialized);
+  auto changesCount = document.select_node("mmwatch/edits/@value").attribute().as_int(-1);
+  auto rank = document.select_node("mmwatch/rank/@value").attribute().as_int(-1);
+
+  lock_guard<mutex> g(m_mutex);
+  if (m_userName != userName)
+    return false;
+
+  m_lastUpdate = time(nullptr);
+  m_userStats = make_shared<UserStats>(m_lastUpdate, rank, changesCount);
+  SaveToSettings();
+
+  return true;
+}
+
+void UserStatsLoader::Update(string const & userName, TOnUpdateColback const & fn)
+{
+  {
+    lock_guard<mutex> g(m_mutex);
+    auto const nothingToUpdate = m_userStats && m_userName == userName && m_userStats &&
+                                 difftime(m_lastUpdate, time(nullptr)) < kOneHour;
+
+    if (nothingToUpdate)
+    {
+      GetPlatform().RunOnGuiThread(fn);
+      return;
+    }
+  }
+
+  thread([this, userName, &fn] {
+    if (Update(userName))
+      GetPlatform().RunOnGuiThread(fn);
+  }).detach();
```

Comment:
Detach in future destructor is guarantied only for c++14 
http://en.cppreference.com/w/cpp/thread/future/%7Efuture